OSOE-609: Add fake video capture sample test in Lombiq.UITestingToolbox

dministro opened the pull request:
[OSOE-609](https://lombiq.atlassian.net/browse/OSOE-609)

[OSOE-609]: https://lombiq.atlassian.net/browse/OSOE-609?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

dministro committed:
Adding sample application and recipe

dministro committed:
Fix spelling

dministro committed:
Adding UI tests

dministro committed:
Adding comments

0liver commented on Lombiq.VueJs.Samples/Views/ContentPart.Edit-BusinessCard.cshtml:
```diff
@@ -0,0 +1,11 @@
+
+<div class="mb-3">
+    <p>@T["QR code:"]</p>
+    <qr-code content="@(Json.Serialize(new { CardId = Model.ContentPart.ContentItem.ContentItemId}))" class="mx-auto d-block"></qr-code>
```
The parentheses can be omitted:
```suggestion
    <qr-code content="@Json.Serialize(new { CardId = Model.ContentPart.ContentItem.ContentItemId})" class="mx-auto d-block"></qr-code>
```

0liver commented on Lombiq.VueJs.Samples/ResourceManagementOptionsConfiguration.cs:
```diff
@@ -30,6 +30,15 @@ static ResourceManagementOptionsConfiguration()
+        _manifest
+            .DefineSingleFileComponent(QrCard)
+            .SetDependencies(LoadingIndicator)
+            .SetDependencies(BusinessCard);
```
This can be a single line:
```suggestion
            .SetDependencies(LoadingIndicator, BusinessCard);
```
![image](https://user-images.githubusercontent.com/164360/232741749-34027319-4725-4a69-9e74-d0c11b36e224.png)

0liver commented on Lombiq.VueJs.Samples/Recipes/Lombiq.VueJs.Samples.Content.recipe.json:
```diff
@@ -0,0 +1,96 @@
+      "name": "content",
+      "data": [
+        {
+          "ContentItemId": "498bwtc3jjqm4w8drpx7kc16xh",
```
Shouldn't we use `[js: uuid()]` here and below?

dministro replied:
No, this is required because the ContentItemId is encoded in the QR code in test videos.

I can use `[js: uuid()]` only for `Jane, Doe`

0liver commented on Lombiq.VueJs.Samples/Lombiq.VueJs.Samples.csproj:
```diff
@@ -33,14 +33,17 @@
     <None Remove="node_modules\**" />
   </ItemGroup>
 
+
```
Remove extraneous empty line.

0liver commented on Lombiq.VueJs.Samples/Controllers/QrCardController.cs:
```diff
@@ -0,0 +1,41 @@
+    // Open this from under /Lombiq.VueJs.Samples/QrCard/Index.
+    public ActionResult Index()
+    {
+        var linkGenerator = HttpContext.RequestServices.GetRequiredService<LinkGenerator>();
```
Looks like a lazy dependency to me, isn't it?

dministro replied:
Can you explain, please?

0liver replied:
Why don't you use regular constructor injection for `LinkGenerator`?

dministro replied:
Good question. I'll do that.

0liver replied:
I thought you didn't do that in the first place because only one of two methods on the class use this dependency, that's why I suggested to use a `Lazy<T>` dependency.

dministro replied:
I'm not sure I can use `Lazy<T>` here, but I'll try it.

dministro replied:
There is no `Lazy<LinkGenerator>` in the service container, but I added `LinkGenerator` as constructor dependency.

0liver replied:
If **Autofac** is used, `Lazy<T>` resolves automatically. But is it?

dministro replied:
No, this is [`Microsoft.Extensions.DependencyInjection.ServiceCollection`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.servicecollection?view=dotnet-plat-ext-7.0).

https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines

0liver replied:
😏 OK! Thanks for the insight.

sarahelsaig replied:
For MS DI we have an extension method in _Lombiq.HelpfulLibraries.Common_ that enables Lazy injection. [See docs here.](https://github.com/Lombiq/Helpful-Libraries/blob/80c5e6c9ac65a0e1c58cb4413ae12228edd59d35/Lombiq.HelpfulLibraries.Common/Docs/DependencyInjection.md#lazy-injection-support)

dministro replied:
Good to know, thank you.

0liver replied:
Wow! So much magic hidden away ✨

0liver commented on Lombiq.VueJs.Samples/Controllers/QrCardController.cs:
```diff
@@ -0,0 +1,41 @@
+        });
+    }
+
+    public async Task<ActionResult> GetBusinessCard(string cardId)
+    {
+        var businessCard = await _contentManager.GetAsync(cardId);
+        if (businessCard is null)
+        {
+            return NotFound();
+        }
+
+        return Ok(businessCard);
+    }
```
Syntactic sugar, extra sweet:
```suggestion
    public async Task<ActionResult> GetBusinessCard(string cardId) =>
        await _contentManager.GetAsync(cardId) is { } businessCard
            ? Ok(businessCard)
            : NotFound();
```

0liver commented on Lombiq.VueJs.Samples/Controllers/QrCardController.cs:
```diff
@@ -0,0 +1,41 @@
+    public QrCardController(IContentManager contentManager) =>
+        _contentManager = contentManager;
+
+    // Open this from under /Lombiq.VueJs.Samples/QrCard/Index.
```
Make this an XmlDoc comment, so it can be seen on hover where it's used, too:
```suggestion
    /// <remarks><para>Open this from under /Lombiq.VueJs.Samples/QrCard/Index.</para></remarks>
```

0liver commented on Lombiq.VueJs.Samples/Views/QrCard/Index.cshtml:
```diff
@@ -0,0 +1,23 @@
+                InsecureContextError: '@T["Camera access is only permitted in secure context. Use HTTPS or localhost rather than HTTP."]',
+                Unknown: '@T["Camera error."]',
+        }"
+        api-url="@Model.ApiUrl"></qr-card>
```
What would you say about wrapping the closing tag to a new line for better visibility?
```suggestion
        api-url="@Model.ApiUrl">
    </qr-card>
```

0liver replied:
In that case, do the same in _[ContentPart.Edit-BusinessCard.cshtml](https://github.com/Lombiq/Orchard-Vue.js/pull/81/files#diff-abc8774add695e5a8f7451b7d9dd57ad6a054599ce9cd2452104cde144cbfb6fR10)_, too, please.

0liver commented on Lombiq.VueJs.Tests.UI/Extensions/BrowserConfigurationExtensions.cs:
```diff
@@ -0,0 +1,24 @@
+    private static void ConfigureFakeVideoSource(this BrowserConfiguration browserConfiguration, string resource) =>
+        browserConfiguration.FakeVideoSource = new FakeBrowserVideoSource
+        {
+            StreamProvider = () =>
+                typeof(BrowserConfigurationExtensions).Assembly.GetManifestResourceStream(
+                    typeof(DirectoryPlaceholder),
+                    resource),
```
I would use the same type for accessing the assembly and the resource stream, just to be extra safe that we're not trying to access a resource from a different assembly:
```suggestion
            StreamProvider = () => typeof(DirectoryPlaceholder)
                .Assembly
                .GetManifestResourceStream(typeof(DirectoryPlaceholder), resource),
```

0liver commented on Lombiq.VueJs.Tests.UI/Extensions/BrowserConfigurationExtensions.cs:
```diff
@@ -0,0 +1,24 @@
+public static class BrowserConfigurationExtensions
+{
+    public static void ConfigureFakeVideoSourceForNegativeTest(this BrowserConfiguration browserConfiguration) =>
+        browserConfiguration.ConfigureFakeVideoSource($"richard.roe.mjpeg");
+
+    public static void ConfigureFakeVideoSourceForPositiveTest(this BrowserConfiguration browserConfiguration) =>
+        browserConfiguration.ConfigureFakeVideoSource($"john.doe.mjpeg");
```
No need for string interpolation:
```suggestion
        browserConfiguration.ConfigureFakeVideoSource("richard.roe.mjpeg");

    public static void ConfigureFakeVideoSourceForPositiveTest(this BrowserConfiguration browserConfiguration) =>
        browserConfiguration.ConfigureFakeVideoSource("john.doe.mjpeg");
```

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/qr-card.vue:
```diff
@@ -0,0 +1,121 @@
+                    NotAllowedError: 'You need to grant camera access permission.',
+                    NotFoundError: 'No camera found.',
+                    NotSupportedError: 'Secure context required (HTTPS, localhost).',
+                    NotReadableError: 'The camera already in use.',
```
```suggestion
                    NotReadableError: 'The camera is already in use.',
```

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/business-card.vue:
```diff
@@ -0,0 +1,74 @@
+            <p v-if="currentCard.BusinessCard.Email.Text">Email: {{ currentCard.BusinessCard.Email.Text }}</p>
+        </div>
+        <div class="card-body" v-else-if="error">
+            <h5>Something went wrong :(</h5>
```
Localize and rephrase:
```suggestion
            <h5>T["An error occurred."]</h5>
```

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/qr-card.vue:
```diff
@@ -0,0 +1,121 @@
+                return {
+                    ...messages,
+                    ...(raw || {}),
+                };
+            },
+        },
+        apiUrl: {
+            type: String,
+            required: true,
+        },
+    },
+    data: function () {
+        return {
+            cardId: null,
+            error: '',
+            loading: false,
+        };
+    },
+    methods: {
+        onDecode(result) {
+            try {
+                const content = JSON.parse(result);
+
+                if (content.cardId) {
+                    this.cardId = content.cardId;
+                }
+            }
+            catch {
+                // Nothing to do here, just continue scanning.
+            }
+        },
```
I expected the app to decode my QR code in this method, but my breakpoint in line 75 was never hit while holding my phone in front of the camera:

![Uploading image.png…]()

dministro replied:
@0liver I can't see your image:
![image](https://user-images.githubusercontent.com/15349881/232778841-5301ceb3-3a8d-4784-82ba-a80fde40d558.png)

0liver replied:
![2023-04-18_12h48_07](https://user-images.githubusercontent.com/164360/232783688-6ba59c58-3b4c-4b49-a3b3-3d3ea15906fa.jpg)

0liver replied:
I tried this both in Opera and Vivaldi, which are both Chromium-based browsers.

dministro replied:
Maybe it can be a problem, if you scan the code before placing a breakpoint, because the app restricts scanning the same code repeatedly.

Chrome:
![image](https://user-images.githubusercontent.com/15349881/232790556-e31646a4-7438-41e4-a9be-7608bc18b355.png)

Firefox:
![image](https://user-images.githubusercontent.com/15349881/232792334-2094ecf3-1c6c-49d3-97dd-b8864945d1e5.png)

0liver replied:
>  if you scan the code before placing a breakpoint, because the app restricts scanning the same code repeatedly.

That definitely helped 😉

So, I've hit the breakpoint now, but after continuing I see no feedback on my screen. Since I'm not scanning a business card but some other QR code I would expect to see an error message of the some sort, e.g. "An error occurred. The given QR code does not contain the expected information."

Did you anticipate that?

dministro replied:
This is why the `try...catch` block. I won't over do this. This is a really minimal implementation of a virtual business card service. All foreseen errors are handled, so I think this UX problem is not an issue.

0liver replied:
I'm just suggesting adding a simple error message like this when you scan a non-business-card QR code:

![image](https://user-images.githubusercontent.com/164360/233015408-86e72c49-fa35-4ae1-afbe-d7602cd1bffc.png)

```suggestion
        onDecode(result) {
            try {
                const content = JSON.parse(result);

                if (content.cardId) {
                    this.cardId = content.cardId;
                    this.error = '';
                }
            }
            catch {
                // Let the user know about incompatible QR codes, and continue scanning.
                this.error = 'This QR code does not represent a business card.';
            }
        },
```

dministro replied:
Ok, please set the issue back to to-do.

dministro committed:
Apply suggestions from code review

Co-authored-by: 0liver <0liver@users.noreply.github.com>

dministro committed:
Addressing "In that case, do the same in ContentPart.Edit-BusinessCard.cshtml, too, please."
Using [js: uuid()] for Jane, Doe
Addressing "Remove extraneous empty line."
Adding LinkGenerator as a construcotr dependency
Localizing everything

dministro committed:
Using Lazy

dministro committed:
Better error handling

0liver commented on Lombiq.VueJs.Samples/Views/QrCard/Index.cshtml:
```diff
@@ -11,6 +11,7 @@
             OverconstrainedError: '@T["Installed cameras are not suitable."]',
             StreamApiNotSupportedError: '@T["The stream API is not supported in this browser."]',
             InsecureContextError: '@T["Camera access is only permitted in secure context. Use HTTPS or localhost rather than HTTP."]',
+            InvalidCard: '@T["This QR code does not represent a business card."]',
```
Do I understand correctly that you're passing these localized messages as overrides for the identical, non-localized messages in [qr-card.vue](https://github.com/Lombiq/Orchard-Vue.js/pull/81/commits/2c047ac05a2f9dbd8c1bd0fe440309e3e4f04728#diff-6b13f557f039659c4a7f222b34a5313f936aa61a0200aaf3ecc1a10a365baa22R56)? Because those messages do look a lot like duplicates 😁

dministro replied:
Yes. I can't use `IViewLocalizer` in Vue app.

0liver replied:
You [**can** use](https://github.com/Lombiq/Orchard-Vue.js/blob/dev/Readme.md#using-vuejs-single-file-components) `IStringLocalizer`, but you would have to move the messages to the `<template>` somehow. Would that make sense?

dministro replied:
Currently, I don't see how, but in that case, don't we need defaults in the code?
```javascript
        errorMessages: {
            type: Object,
            default: function (raw) {
                const messages = {
                    NotAllowedError: 'You need to grant camera access permission.',
                    NotFoundError: 'No camera found.',
                    NotSupportedError: 'Secure context required (HTTPS, localhost).',
                    NotReadableError: 'The camera is already in use.',
                    OverconstrainedError: 'Installed cameras are not suitable.',
                    StreamApiNotSupportedError: 'The stream API is not supported in this browser.',
                    InsecureContextError: 'Camera access is only permitted in secure context. Use HTTPS or localhost rather than HTTP.',
                    InvalidCard: 'This QR code does not represent a business card.',
                    Unknown: 'Camera error.',
                };

                return {
                    ...messages,
                    ...(raw || {}),
                };
            },
        },
```

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/qr-card.vue:
```diff
@@ -83,20 +89,29 @@ export default {
+            })(result);
+
+            if (!content.cardId) {
+                this.cardId = '';
```
For consistency, set `cardId` to `null` just like in [line 89](https://github.com/Lombiq/Orchard-Vue.js/pull/81/commits/2c047ac05a2f9dbd8c1bd0fe440309e3e4f04728#diff-6b13f557f039659c4a7f222b34a5313f936aa61a0200aaf3ecc1a10a365baa22R89).

0liver reviewed (changes requested):
Much nicer than my suggestion!

dministro committed:
Addressing "For consistency, set cardId to null just like in line 89."

dministro committed:
Localizing strings in template

dministro committed:
Removing message property of loading-indicator

dministro committed:
Fixing analizer violation

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/business-card.vue:
```diff
@@ -0,0 +1,74 @@
+                const response = await fetch(`${this.apiUrl}?cardId=${newId}`);
+
+                if (!response.ok) {
+                    const message = `An error has occurred: ${response.status} - ${response.statusText}`;
+
+                    throw new Error(message);
```
I've inlined this variable.

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/business-card.vue:
```diff
@@ -0,0 +1,74 @@
+<template>
+    <div class="card qr-card" v-if="!loading && cardId">
+        <div class="card-body" v-if="!error && currentCard">
+            <h5 class="full-name">{{ currentCard.BusinessCard.FirstName.Text }}, {{ currentCard.BusinessCard.LastName.Text }}</h5>
```
I've removed the comma between first and last name.

dministro replied:
Let me adjust the test here:
https://github.com/Lombiq/Orchard-Vue.js/blob/a3c4b5265e6e7c9546f9c975ccb667695cacd577/Lombiq.VueJs.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs#L110-L116

0liver replied:
Thanks! I didn't catch that 😏

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/qr-card.vue:
```diff
@@ -0,0 +1,124 @@
+    },
+    methods: {
+        onDecode(result) {
+            const content = ((text) => {
+                try {
+                    return JSON.parse(text);
+                }
+                catch {
+                    return {};
+                }
+            })(result);
```
It's sad that this can't be compacted into a single line, but the idea is great 😁

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/qr-card.vue:
```diff
@@ -0,0 +1,124 @@
+                }
+            })(result);
+
+            if (!content.cardId) {
+                this.cardId = null;
+                this.scanError = 'InvalidCard';
+
+                return;
+            }
+
+            this.cardId = content.cardId;
+            this.scanError = '';
```
How about saving 7 lines?
```suggestion
            this.cardId = content.cardId;
            this.scanError = content.cardId ? '' : 'InvalidCard';
```

0liver replied:
@dministro Just commit if you agree, or reply here - I'll do the rest.

0liver commented on Lombiq.VueJs.Samples/Views/QrCard/Index.cshtml:
@DAud-IcI I wonder if this can be even more Vue-ified? I've looked at https://github.com/Lombiq/Orchard-Vue.js/blob/dev/Readme.md#using-vuejs-single-file-components, but I don't quite see it. Still, I'm under the impression that this view is just boilerplate code that you've already addressed somehow.

0liver replied:
@dministro Let's not wait for Dávid's input here - I'd just like to know.

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/business-card.vue:
```diff
@@ -0,0 +1,74 @@
+    <div class="card qr-card" v-if="!loading && cardId">
+        <div class="card-body" v-if="!error && currentCard">
+            <h5 class="full-name">{{ currentCard.BusinessCard.FirstName.Text }}, {{ currentCard.BusinessCard.LastName.Text }}</h5>
+            <p v-if="currentCard.BusinessCard.Phone.Text">[[ Phone ]]: {{ currentCard.BusinessCard.Phone.Text }}</p>
+            <p v-if="currentCard.BusinessCard.Email.Text">[[ Email ]]: {{ currentCard.BusinessCard.Email.Text }}</p>
```
🥳 for using inline localization!

Commit:
Remove comma, add line, inline variable

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/qr-card.vue:
```diff
@@ -42,47 +49,10 @@ export default {
-                    ...(raw || {}),
-                };
-            },
-        },
```
Sooo nice!

dministro committed:
Update Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/qr-card.vue

Co-authored-by: 0liver <0liver@users.noreply.github.com>

dministro committed:
Adjusting UI test

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/business-card.vue:
```diff
@@ -0,0 +1,73 @@
+        </div>
+        <div class="card-body" v-else-if="error">
+            <h5>[[ An error occurred. ]]</h5>
+            <p class="error-message">{{ error }}</p>
```
The current class doesn't add any styling - let's use `message-error` from BS, and `px-1` for some horizontal padding:
```suggestion
            <p class="message-error px-1">{{ error }}</p>
```

0liver replied:
See screenshots here for the result: https://github.com/Lombiq/Orchard-Vue.js/pull/81#discussion_r1175036860.

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/qr-card.vue:
```diff
@@ -0,0 +1,117 @@
+        </div>
+        <div class="row justify-content-center mb-3" v-if="error">
+            <div class="col-md-6 col-sm">
+                <p v-if="error === 'NotAllowedError'" class="error">[[ You need to grant camera access permission. ]]</p>
+                <p v-else-if="error === 'NotFoundError'" class="error">[[ No camera found. ]]</p>
+                <p v-else-if="error === 'NotSupportedError'" class="error">[[ Secure context required (HTTPS, localhost). ]]</p>
+                <p v-else-if="error === 'NotReadableError'" class="error">[[ The camera is already in use. ]]</p>
+                <p v-else-if="error === 'OverconstrainedError'" class="error">[[ Installed cameras are not suitable. ]]</p>
+                <p v-else-if="error === 'StreamApiNotSupportedError'" class="error">[[ The stream API is not supported in this browser. ]]</p>
+                <p v-else-if="error === 'InsecureContextError'" class="error">[[ Camera access is only permitted in secure context. Use HTTPS or localhost rather than HTTP. ]]</p>
+                <p v-else class="error">[[ Camera error. ]]</p>
```
Let's use some working error classes:

```suggestion
                <p v-if="error === 'NotAllowedError'" class="message-error px-1">[[ You need to grant camera access permission. ]]</p>
                <p v-else-if="error === 'NotFoundError'" class="message-error px-1">[[ No camera found. ]]</p>
                <p v-else-if="error === 'NotSupportedError'" class="message-error px-1">[[ Secure context required (HTTPS, localhost). ]]</p>
                <p v-else-if="error === 'NotReadableError'" class="message-error px-1">[[ The camera is already in use. ]]</p>
                <p v-else-if="error === 'OverconstrainedError'" class="message-error px-1">[[ Installed cameras are not suitable. ]]</p>
                <p v-else-if="error === 'StreamApiNotSupportedError'" class="message-error px-1">
                    [[ The stream API is not supported in this browser. ]]
                </p>
                <p v-else-if="error === 'InsecureContextError'" class="message-error px-1">
                    [[ Camera access is only permitted in secure context. Use HTTPS or localhost rather than HTTP. ]]
                </p>
                <p v-else class="message-error px-1">[[ Camera error. ]]</p>
```

dministro replied:
I'll add `message-error px-1` to the parent container instead.

dministro replied:
See: https://github.com/Lombiq/Orchard-Vue.js/pull/81/commits/a142ed8b1dc013ddc97d0fea754016112320bb32

0liver commented on Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/qr-card.vue:
```diff
@@ -0,0 +1,117 @@
+        </div>
+        <div class="row justify-content-center mb-3" v-if="scanError">
+            <div class="col-md-6 col-sm">
+                <p v-if="scanError == 'InvalidCard'" class="error">[[ This QR code does not represent a business card. ]]</p>
```
```suggestion
                <p v-if="scanError == 'InvalidCard'" class="message-error px-1">[[ This QR code does not represent a business card. ]]</p>
```

0liver replied:
Before:

![image](https://user-images.githubusercontent.com/164360/233960402-8c8856f8-1a06-42bb-b035-a55c3d88a9ce.png)

After:

![image](https://user-images.githubusercontent.com/164360/233960463-575695b0-7a22-456c-ba20-0551916c7307.png)

dministro committed:
Update Lombiq.VueJs.Samples/Assets/Scripts/VueComponents/business-card.vue

Co-authored-by: 0liver <0liver@users.noreply.github.com>

dministro committed:
Better error styling

dministro committed:
Adjusting UI test

Commit:
Remove unnecessary class
